pynamodb/connection/base.py

```diff
@@ -1384,10 +1386,14 @@ def _reverse_dict(d):
         return {v: k for k, v in d.items()}
 
 
-def _convert_binary(attr):
+def _convert_binary(attr: Dict[str, Any]) -> None:
     if BINARY in attr:
-        attr[BINARY] = b64decode(attr[BINARY].encode(DEFAULT_ENCODING))
+        attr[BINARY] = b64decode(attr[BINARY].encode())
     elif BINARY_SET in attr:
-        value = attr[BINARY_SET]
-        if value and len(value):
-            attr[BINARY_SET] = {b64decode(v.encode(DEFAULT_ENCODING)) for v in value}
+        attr[BINARY_SET] = [b64decode(v.encode()) for v in attr[BINARY_SET]]
+    elif MAP in attr:
+        for sub_attr in attr[MAP].values():
+            _convert_binary(sub_attr)
+    elif LIST in attr:
+        for sub_attr in attr[LIST]:
+            _convert_binary(sub_attr)
```

Comment:
Maps and lists are now being recursed into. This matches botocore behavior.

Reply:
Worth noting that this is going to have some performance impact (I actually did notice this as part of the delta in https://github.com/pynamodb/PynamoDB/pull/1079). We might want to run benchmarks to get a sense for it and include it in the release notes.

If the difference is meaningful then we may also want to add an optimisation for short-circuiting this when binary attributes aren't in use